arelle/plugin/validate/NL/rules/fg_nl.py

```diff
@@ -0,0 +1,90 @@
+"""
+See COPYRIGHT.md for copyright information.
+"""
+from __future__ import annotations
+
+from typing import Any, Iterable
+
+from arelle import ModelDocument
+from arelle.ModelInstanceObject import ModelFact
+from arelle.ModelObject import ModelObject
+from arelle.ModelValue import qname
+from arelle.ValidateXbrl import ValidateXbrl
+from arelle.XbrlConst import qnLinkSchemaRef, qnLinkLinkbaseRef, qnLinkRoleRef, qnLinkArcroleRef, qnXbrliContext, qnXbrliUnit
+from arelle.typing import TypeGetText
+from arelle.utils.PluginHooks import ValidationHook
+from arelle.utils.validate.Decorator import validation
+from arelle.utils.validate.Validation import Validation
+from ..DisclosureSystems import (
+    DISCLOSURE_SYSTEM_NT16,
+    DISCLOSURE_SYSTEM_NT17,
+    DISCLOSURE_SYSTEM_NT18,
+)
+from ..PluginValidationDataExtension import PluginValidationDataExtension
+
+_: TypeGetText
+
+INSTANCE_ELEMENT_ORDER = {
+    qnLinkSchemaRef.clarkNotation: 0,
+    qnLinkLinkbaseRef.clarkNotation: 1,
+    qnLinkRoleRef.clarkNotation: 2,
+    qnLinkArcroleRef.clarkNotation: 3,
+    qnXbrliContext.clarkNotation: 4,
+    qnXbrliUnit.clarkNotation: 5,
+    'fact': 6,
+}
+
+
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+    disclosureSystems=[
+        DISCLOSURE_SYSTEM_NT16,
+        DISCLOSURE_SYSTEM_NT17,
+        DISCLOSURE_SYSTEM_NT18,
+    ],
+)
+def rule_fg_nl_04(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> Iterable[Validation] | None:
+    """
+    FG-NL-04: An XBRL instance document SHOULD order the elements so that referents precede references.
+    The children of the 'xbrli:xbrl' element should appear in the following order:
+    1. schemaRef
+    2. linkbaseRef
+    3. roleRef
+    4. arcroleRef
+    5. context
+    6. units
+    7. facts
+    """
+    modelXbrl = val.modelXbrl
+    for doc in modelXbrl.urlDocs.values():
+        if doc.type == ModelDocument.Type.INSTANCE:
+            currentName = list(INSTANCE_ELEMENT_ORDER.keys())[0]
+            currentIndex = INSTANCE_ELEMENT_ORDER[currentName]
+            for elt in doc.xmlRootElement.iter():
+                if not isinstance(elt, ModelObject):
+                    continue
+                if isinstance(elt, ModelFact):
+                    thisName = 'fact'
+                else:
+                    thisName = elt.elementQname.clarkNotation
+                if thisName not in INSTANCE_ELEMENT_ORDER:
+                    continue
+                thisIndex = INSTANCE_ELEMENT_ORDER[thisName]
+                if thisIndex == currentIndex:
+                    continue
+                if thisIndex > currentIndex:
+                    currentIndex = thisIndex
+                    currentName = thisName
+                else:
+                    yield Validation.warning(
```

Comment:
I think that would have the effect of masking other issues with the ordering. For example, in the included test case only the `"context" after "fact"` error would show since the "current" type is now "context". The absence of a `"unit" after "fact"` message may lead someone to hastily move the fact between "context" and "unit" which would cause another error on the next run. I lean towards more information upfront.

I do prefer consolidating to fewer warnings with multiple refs, so I'll make an update for that.